### PR TITLE
[WIP] support setting the parent for organization accounts

### PR DESCRIPTION
**Pull request by afeld**
Part of #571. Builds on #4207 - [diff](https://github.com/afeld/terraform-provider-aws/compare/ou...account-parent).

### CODE REVIEW

**Comment by afeld** [comment]
Getting some errors I don't understand, and could use another set of eyes on this:

```
$ TEST_AWS_ORGANIZATION_ACCOUNT_EMAIL_DOMAIN=gmail.com make testacc TEST=./aws TESTARGS="-run TestAccAWSOrganizations/Account/basic"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run TestAccAWSOrganizations/Account/basic -timeout 120m
=== RUN   TestAccAWSOrganizations
=== RUN   TestAccAWSOrganizations/Account
=== RUN   TestAccAWSOrganizations/Account/basic
--- FAIL: TestAccAWSOrganizations (18.84s)
    --- FAIL: TestAccAWSOrganizations/Account (18.83s)
        --- FAIL: TestAccAWSOrganizations/Account/basic (18.83s)
        	testing.go:518: Step 0 error: Check failed: Check 4/8 error: aws_organizations_account.test: Attribute 'joined_timestamp' expected to be set
        	testing.go:579: Error destroying resource! WARNING: Dangling resources
        		may exist. The full state and error is shown below.

        		Error: Error applying: 1 error(s) occurred:

        		* aws_organizations_account.test (destroy): 1 error(s) occurred:

        		* aws_organizations_account.test: ConstraintViolationException: The member account must be configured with a payment method, such as a credit card.
        			status code: 400, request id: XXXXXXX

        		State: aws_organizations_account.test:
        		  ID = XXXXXXX
        		  provider = provider.aws
        		  arn = arn:aws:organizations::XXXXXXXX:account/o-XXXXXXX/XXXXXXXX
        		  email = tf-acctest+XXXXXXXXXX@gmail.com
        		  joined_method = CREATED
        		  name = tf_acctest_XXXXXXXXXX
        		  parent_id = r-XXXX
        		  status = ACTIVE
```

> `Attribute 'joined_timestamp' expected to be set`

Why would this be missing, of all things?

> `Error destroying resource!`

The error is what I would expect if Terraform were trying to delete the account, but this error isn't happening on other branches, so somehow I'm causing a change in the cleanup behavior. Anyone have ideas?

**Comment by afeld** [comment]
Correction: I _am_ seeing those same errors on `master`. @asedge Is there a trick to running these tests?

**Comment by asedge** [comment]
@afeld This is the error:
```
ConstraintViolationException: The member account must be configured with a payment method, such as a credit card.
```
AWS Organization member accounts are a real pain to try to clean up.

**Comment by asedge** [comment]
Here was the discussion around trying to do the delete: https://github.com/terraform-providers/terraform-provider-aws/pull/3524#discussion_r170692979. Prior, I just had printed a message that said you need to do the delete from the Web UI or something along those lines - which wasn't a great solution.

**Comment by bflad** [comment]
This resource is a real pain to acceptance test due to how Organizations/AWS accounts in general work. Basically, its not really doable in an automated fashion since AWS requires converting accounts to standalone (requiring EULA agreement, credit card information, phone verification, etc.) and there is no simple `DeleteAccount` API call.

I'm not sure we want to implement a flag attribute to leave the account dangling in the organization during resource deletion just for the acceptance testing either (also doesn't seem like a valid regular use case when there is `state rm`). 🙁 

We could switch this to `service/organizations/iface` for the acceptance testing, but that also has its own problems with keeping it appropriately up to date.

In short, I'm not sure there is a good way to handle testing for this resource 😅 -- I wanted to get the resource merged as it worked as expected with manual testing. I would highly recommend opening an AWS support case if you have a support plan that asks for more automated ways to handle deleting an account (either the missing DeleteAccount API call or APIs for the billing/EULA pieces. Personally, I'd accept a PR with manual testing for now. 👍

**Comment by breser** [comment]
I have Enterprise Support at work and I'm starting to use this stuff for work.  So I'll start a discussion with them around this.

**Comment by afeld** [comment]
@asedge Right, the error makes sense to me... Are you saying that the `basic` acceptance test is known to fail on `master`?

> I have Enterprise Support at work and I'm starting to use this stuff for work. So I'll start a discussion with them around this.

Yep, I already have as well, but couldn't hurt for them to hear it again. They are understandably reluctant to make it possible to delete all of your infrastructure with one API call.

**Comment by breser** [comment]
@afeld They added us to the list of ~40 customers that have asked for this.

**Comment by afeld** [comment]
Repeating my question from above: is the `basic` acceptance test is known to fail on `master`?

**Comment by bflad** [comment]
> Repeating my question from above: is the basic acceptance test is known to fail on master?

Yes, see my [comment](https://github.com/terraform-providers/terraform-provider-aws/pull/4405#issuecomment-385716972) above about the various issues encountered trying to implement the automated testing. I would accept a manually tested PR in this rare scenario. Adding automated tests with `t.Skip()` would be nice as well in case there is a good workaround in the future.

**Comment by NissesSenap** [comment]
Any new status on this pull request? 
I can understand all the billing stuff is a pain, would a possible way forward be to split up the easier stuff like aws_organizations_unit be split up to a separate pull request so it gets merged?

**Comment by afeld** [comment]
> split up the easier stuff like aws_organizations_unit

See #4207.

**Comment by bflad** [comment]
@afeld would you be okay if I cherry-picked the following two of your commits into a new pull request? 

* https://github.com/terraform-providers/terraform-provider-aws/pull/4405/commits/d7f007c42392d24fc1b388e95787fc387c0c29ff
* https://github.com/terraform-providers/terraform-provider-aws/pull/4405/commits/3b071fdcc1aff50f4ce01771098854a856fcea74

Now that we have Organizational Unit support would love to use your work as a starting point for this enhancement without the messy rebase. Thanks!

**Comment by afeld** [comment]
Go for it!

**Comment by bflad** [comment]
Thanks so much, @afeld! Including those two commits as the start, I was able to finish the implementation here: #8583

Since the new pull request now supersedes this one, closing this. Please follow that new pull request (or its tracking feature request #8281) for further updates. 👍
